### PR TITLE
Save empty DONE file to S3 bucket when db dump is done

### PR DESCRIPTION
**Pull request by YinfangZ**

### CODE REVIEW

File: lib/aurora_bootstrapper/table.rb [hunk 1]
```diff
@@ -54,9 +58,30 @@ def json_object
...
       @client.query( export_statement( into_bucket: into_bucket ) )
-      AuroraBootstrapper.logger.info( message:"Export succeeded: #{@database_name}.#{@table_name}" )
+      AuroraBootstrapper.logger.info( message: "Export succeeded: #{@database_name}.#{@table_name}" )
+
+      # export state to S3
+      index = into_bucket.rindex('/')
+      if index > 5
+        into_bucket = into_bucket[5..index-1]
+      else
+        into_bucket = into_bucket[5..-1]
```
Comment (gaorlov) [inline lib/aurora_bootstrapper/table.rb:70]: what happens if the bucket pathing changes?

File: lib/aurora_bootstrapper/table.rb [hunk 2]
```diff
@@ -73,5 +98,21 @@ def export_statement( into_bucket: )
           OVERWRITE ON
       SQL
     end
+
+    def object_uploaded?(s3_client, bucket_name, object_key)
+      response = s3_client.put_object(
```
Comment (gaorlov) [inline lib/aurora_bootstrapper/table.rb:103]: it is surprising to me as a reader to see a side effect in a `*?` method. Can we move this into a class? like a 
```ruby 
class Notifier
  def initialize( client )
    @client = client
  end

  def push_state( table )
    @client.whagever #=> push update about table
  end
end
```

**Review by gaorlov — CHANGES_REQUESTED** [review]
This is a great start! I'm excited to see notifiers for the exports come to life! 
I have a few questions:
* Am I correct in remembering that we are pushing to S3 to trigger an message? Can we push to the message bus directly?
* Can we move notifications into a new class? It will make testing more isolated.

Let me know if you want to chat about any of these. I'm excited to see this ship!

File: bin/exporter [hunk 3]
```diff
@@ -23,9 +24,12 @@ client    = ConnectionPool::Wrapper.new(size: 15, timeout: 3) do
...
+
 AuroraBootstrapper::Exporter.new( client: client,
                                   prefix: prefix,
                            export_bucket: bukkit,
                       blacklisted_tables: blacklisted_tables,
                       whitelisted_tables: whitelisted_tables,
-                      blacklisted_fields: blacklisted_fields ).export!
\ No newline at end of file
+                      blacklisted_fields: blacklisted_fields,
+                      s3_client: s3_client ).export!
```
Comment (gaorlov) [inline bin/exporter:35]: if the client never changes, what's the benefit of passing it in here? Can it be created on demand?

File: lib/aurora_bootstrapper/table.rb [hunk 4]
```diff
@@ -1,11 +1,16 @@
...
-    def initialize( database_name:, table_name:, client:, blacklisted_fields: [] )
+    def initialize( database_name:, table_name:, client:, blacklisted_fields: [], s3_client: nil )
       @blacklisted_fields = blacklisted_fields
       @database_name      = database_name
       @table_name         = table_name
       @client             = client
-      @export_date        = ENV.fetch( 'EXPORT_DATE', nil )
+      d                   = DateTime.now
+      d_str               = d.strftime("%Y-%m-%d")
+      @export_date        = ENV.fetch( 'EXPORT_DATE', d_str )
```
Comment (gaorlov) [inline lib/aurora_bootstrapper/table.rb:12]: are these variables used anywhere?

File: lib/aurora_bootstrapper/table.rb [hunk 5]
```diff
@@ -1,11 +1,16 @@
...
+      @notifier           = Notifier.new s3_client: s3_client
```
Comment (gaorlov) [inline lib/aurora_bootstrapper/table.rb:13]: does this have to come from the outside or can this be defined inside the class?
```ruby
class Table
  def notifier
    @notifier||= Notifier.new
  end
end

class Notifier
  def initialize
    region  = ENV.fetch( 'REGION', 'us-west-2')
    @client = Aws::S3::Client.new(region: region)
  end
end
```

File: bin/exporter (hunk 3 above)
Comment (YinfangZ) [reply bin/exporter:35]: Sure.

File: lib/aurora_bootstrapper/table.rb (hunk 4 above)
Comment (YinfangZ) [reply lib/aurora_bootstrapper/table.rb:12]: At line 12.

File: lib/aurora_bootstrapper/table.rb [hunk 6]
```diff
@@ -54,10 +58,11 @@ def json_object
...
 
     def export!( into_bucket: )
-      AuroraBootstrapper.logger.info( message:"Running: #{ export_statement( into_bucket: into_bucket ) }" )
+      AuroraBootstrapper.logger.info( message: "Running Export: #{ export_statement( into_bucket: into_bucket ) }" )
       @client.query( export_statement( into_bucket: into_bucket ) )
-      AuroraBootstrapper.logger.info( message:"Export succeeded: #{@database_name}.#{@table_name}" )
-      true
+      AuroraBootstrapper.logger.info( message: "Export succeeded: #{@database_name}.#{@table_name}" )
+
+      @notifier.push_state?( into_bucket: into_bucket )
```
[the remark this answers is not in the document]
Comment (YinfangZ) [reply lib/aurora_bootstrapper/table.rb:65]: side-effect method I guess as we output some state to S3.

File: lib/aurora_bootstrapper/table.rb (hunk 4 above)
Comment (YinfangZ) [reply lib/aurora_bootstrapper/table.rb]: I will move this part of code to exporter.rb and pass @export_date around to different other files to make sure we put dump files under the same datetime folder

File: lib/aurora_bootstrapper/table.rb (hunk 5 above)
Comment (YinfangZ) [reply lib/aurora_bootstrapper/table.rb]: Only create s3 client inside Notifier file; move the initialization of notifier to exporter file and use notifier instance inside exporter.rb outside of databases.name loop

File: lib/aurora_bootstrapper/notifier.rb [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+module AuroraBootstrapper
+  class Notifier
+    def initialize( s3_client )
+      @s3_client = s3_client
+    end
+
+    def push_state?( into_bucket: )
+
+      # remove the last subfolder which contains db partition name and the prefix s3://
```
[the remark this answers is not in the document]
Comment (YinfangZ) [reply lib/aurora_bootstrapper/notifier.rb]: Just realized exporter/exporter.rb only covers one exporter; there is no need to put DONE.txt in the parent folder; just put DONE.txt in the current folder

File: lib/aurora_bootstrapper/table.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (YinfangZ) [reply lib/aurora_bootstrapper/table.rb]: How about keep the original name push_state?

File: test/sql/user.sql [hunk 8]
```diff
@@ -1,4 +1,4 @@
-CREATE USER 'root'@'%' IDENTIFIED BY 'root';
+-- CREATE USER 'root'@'%' IDENTIFIED BY 'root';
```
Comment (coding-paras) [inline test/sql/user.sql:1]: Finally, this did the trick :)

Comment (coding-paras) [reply test/sql/user.sql:1]: > Finally, this did the trick :)

What are your thoughts on doing `CREATE USER IF NOT EXISTS ...` rather than commenting?

File: bin/exporter [hunk 9]
```diff
@@ -28,4 +28,5 @@ AuroraBootstrapper::Exporter.new( client: client,
                            export_bucket: bukkit,
                       blacklisted_tables: blacklisted_tables,
                       whitelisted_tables: whitelisted_tables,
-                      blacklisted_fields: blacklisted_fields ).export!
\ No newline at end of file
+                      blacklisted_fields: blacklisted_fields,
+                      notifier: Notifier.new ).export!
```
Comment (gaorlov) [inline bin/exporter:32]: What's the thinking behind passing this into the Exporter as a param? Since there aren't multiple different notifiers, can we init this in the Exporter?

File: lib/aurora_bootstrapper/exporter.rb [hunk 10]
```diff
@@ -4,30 +4,43 @@ module AuroraBootstrapper
...
       @match              = "#{prefix}.*"
       @export_bucket      = export_bucket
       @blacklisted_tables = blacklisted_tables.split(",")
       @whitelisted_tables = whitelisted_tables.split(",")
       @blacklisted_fields = blacklisted_fields.split(",")
       @client             = client
+      @notifier           = notifier
+      d                   = DateTime.now
+      d_str               = d.strftime("%Y-%m-%d")
+      @export_date        = ENV.fetch( 'EXPORT_DATE', d_str )
```
Comment (gaorlov) [inline lib/aurora_bootstrapper/exporter.rb:17]: since this is only used in Table and comes from an env var, what's the thinking behind passing this through Exporter and Database?

Comment (YinfangZ) [reply lib/aurora_bootstrapper/exporter.rb:17]: Notifier also needs this info

File: lib/aurora_bootstrapper/exporter.rb [hunk 11]
```diff
@@ -4,30 +4,43 @@ module AuroraBootstrapper
...
+                             blacklisted_fields: @blacklisted_fields,
+                             export_date: @export_date 
           database.export! into_bucket: @export_bucket
         rescue => e
           AuroraBootstrapper.logger.error message: "Error in database #{database_name}", error: e
         end
       end
+
+      result = true
+      if @notifier
```
Comment (gaorlov) [inline lib/aurora_bootstrapper/exporter.rb:39]: Do we expect the notifier to ever not be present?

Ruby has implicit returns, so you can do
```ruby
notifier.push_state
```

File: bin/exporter (hunk 9 above)
Comment (YinfangZ) [reply bin/exporter:32]: For testing purpose

Comment (YinfangZ) [reply bin/exporter:32]: If I pass it in I can mock the notifier in the test

File: lib/aurora_bootstrapper/table.rb [hunk 12]
```diff
@@ -1,11 +1,13 @@
...
 module AuroraBootstrapper
   class Table
-    def initialize( database_name:, table_name:, client:, blacklisted_fields: [] )
+    def initialize( database_name:, table_name:, client:, blacklisted_fields: [], export_date: nil)
       @blacklisted_fields = blacklisted_fields
       @database_name      = database_name
       @table_name         = table_name
       @client             = client
-      @export_date        = ENV.fetch( 'EXPORT_DATE', nil )
+      @export_date        = export_date
```
Comment (gaorlov) [inline lib/aurora_bootstrapper/table.rb:10]: is this used?

File: lib/aurora_bootstrapper/exporter.rb (hunk 10 above)
Comment (YinfangZ) [reply lib/aurora_bootstrapper/exporter.rb:17]: I created @export_date in exporter, so that the data files and state file will be under the same datetime subfolder even if the job may run longer than one day

File: lib/aurora_bootstrapper/table.rb (hunk 12 above)
Comment (YinfangZ) [reply lib/aurora_bootstrapper/table.rb:10]: In export_statement method

File: bin/exporter (hunk 9 above)
Comment (YinfangZ) [reply bin/exporter:32]: If I create a Notifier inside exporter, then I have no way to mock this class's method in unit tests

File: lib/aurora_bootstrapper/exporter.rb (hunk 11 above)
Comment (YinfangZ) [reply lib/aurora_bootstrapper/exporter.rb:39]: Always do null check because the world is scary :) . Sure I can change that.

Comment (YinfangZ) [reply lib/aurora_bootstrapper/exporter.rb:39]: A second thought: I want to return true if notifier is null; I guess the default behavior of Ruby is return false if notifier is null?

File: lib/aurora_bootstrapper/exporter.rb [hunk 13]
```diff
@@ -4,30 +4,43 @@ module AuroraBootstrapper
...
+        result = @notifier.push_state?( export_date: @export_date, into_bucket: @export_bucket )
```
Comment (gaorlov) [inline lib/aurora_bootstrapper/exporter.rb:40]: since this is the only place we use export date and Notifier is the class that uses it, can it handle its derivation?
like

```ruby
class Exporter
  def notifier
    @notifier ||= Notifier.new( bucket: @export_bucket )
  end
  
  def export!
    #...
    notifier.notify.success?
  end
end

class Notifier
  def initialize( bucket: nil ); end
  
  def export_date
    @export_date ||= ENV.fetch( 'EXPORT_DATE',
                                 DateTime.now.strftime("%Y-%m-%d") )
  end
end
```

Comment (YinfangZ) [reply lib/aurora_bootstrapper/exporter.rb:40]: but table.rb also uses export_date to construct the s3 path to put data files

File: bin/exporter (hunk 9 above)
Comment (gaorlov) [reply bin/exporter:32]: Is is not considered best practice to design code around testing. How can I help get this addressed?

File: lib/aurora_bootstrapper/exporter.rb (hunk 11 above)
Comment (gaorlov) [reply lib/aurora_bootstrapper/exporter.rb:39]: Can we leverage the [null object pattern](https://en.wikipedia.org/wiki/Null_object_pattern) here for the null case?

File: lib/aurora_bootstrapper/notifier.rb [hunk 14]
```diff
@@ -0,0 +1,46 @@
+require 'aws-sdk-s3'
+
+module AuroraBootstrapper
+  class Notifier
+    def initialize( stub_responses: false )
```
Comment (gaorlov) [inline lib/aurora_bootstrapper/notifier.rb:5]: can we avoid test conditions is prod code?

File: lib/aurora_bootstrapper/exporter.rb [hunk 15]
```diff
@@ -4,30 +4,43 @@ module AuroraBootstrapper
...
+      end
+
+      result
```
Comment (gaorlov) [inline lib/aurora_bootstrapper/exporter.rb:43]: does this result get used anywhere?

File: lib/aurora_bootstrapper/notifier.rb [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    def push_state?( export_date:, into_bucket: )
+
+      # remove the prefix s3://
+      into_bucket = into_bucket[5..-1]
+
+      # append export_date (if there is any) and empty state file DONE.txt
+      path = [into_bucket, export_date, 'DONE.txt' ].compact.join('/')
+      index = path.index('/')
+      bucket_name = path[0, index - 1]
+      object_key = path[index + 1..-1]
```
Comment (gaorlov) [inline lib/aurora_bootstrapper/notifier.rb:19]: am i reading this correctly that we take the bucket, append the date and done, split that back up by the first `/` into bucket name and object?

This is a little hard to track and test.  Can we split this into separate methods, like:
```ruby
class Notifier
  def initialize( s3_path: s3_path)
    @s3_path = s3_path
  end
  
  def notify
    client.put_object(
      bucket: bucket_name,
      key: object_key
    )
    AuroraBootstrapper.logger.info( message: "State file has been uploaded to S3 '#{bucket_name}/#{object_key}'." )
  rescue => e
    AuroraBootstrapper.logger.error( message: "State file failed to upload to S3 '#{bucket_name}/#{object_key}': #{e.messages}." )
  end

  protected

  def region
    @region ||= ENV.fetch( 'REGION', 'us-west-2' )
  end

  def client
    @client ||= Aws::S3::Client.new(region: region)
  end
  
  def bucket
    @bucket  ||=.unprefixed_path.split( '/' ).first
  end
  
  def object_key
    @object_key ||= [ bucket_path, export_date, filename ].join( '/' )
  end
  
  def export_date
    @export_date ||= ENV.fetch( 'EXPORT_DATE',
                                 DateTime.now.strftime("%Y-%m-%d") )
  end

  def filename
    'DONE.txt'
  end

  def unprefixed_path
    @unprefixed_path ||= @s3_path.gsub ("s3://", "" )
  end

  def bucket_path
    @bucket_path ||= ( unprefixed_path.split( '/' ) - [ bucket ] ).join( '/' )
  end
end
```

**Review by gaorlov — CHANGES_REQUESTED** [review]
This is excellent progress! The notifier class is missing tests and is a little hard to follow. Once we have some code coverage for that we should be set.

File: lib/aurora_bootstrapper/exporter.rb (hunk 15 above)
Comment (YinfangZ) [reply lib/aurora_bootstrapper/exporter.rb:43]: Removed

File: lib/aurora_bootstrapper/notifier.rb (hunk 16 above)
Comment (YinfangZ) [reply lib/aurora_bootstrapper/notifier.rb:19]: export_date could return different values as we use now for each invocation. Still like my implementaion of initializaing export_date in the constructor of exporter

Comment (gaorlov) [reply lib/aurora_bootstrapper/notifier.rb:19]: That's fair. Should it be a property of the top level AuroraBootstrap module. Also: what happens when the export has no date?

Comment (YinfangZ) [reply lib/aurora_bootstrapper/notifier.rb:19]: Move @export_date ||= ENV.fetch( 'EXPORT_DATE', DateTime.now.strftime("%Y-%m-%d") ) to initialize method; simply return @export_date in export_date method

File: lib/aurora_bootstrapper/exporter.rb (hunk 13 above)
Comment (YinfangZ) [reply lib/aurora_bootstrapper/exporter.rb:40]: pass along @notifier.export_date all the way from exporter.rb to database.rb to table.rb

File: lib/aurora_bootstrapper/exporter.rb (hunk 11 above)
Comment (YinfangZ) [reply lib/aurora_bootstrapper/exporter.rb:39]: Hmm double check the current code logic; looks like @notifier will never be null as it is initialized within the constructor. So I remove the null check logic. But it is good to know this null object pattern.

File: lib/aurora_bootstrapper/exporter.rb [hunk 17]
```diff
@@ -11,23 +11,32 @@ def initialize( client:, prefix: "", export_bucket:, blacklisted_tables: "", whi
...
       database_names.all? do | database_name |
         begin
           @client.query( "use `#{database_name}`" )
           database = Database.new database_name: database_name,
                                          client: @client,
                              blacklisted_tables: @blacklisted_tables,
                              whitelisted_tables: @whitelisted_tables,
-                             blacklisted_fields: @blacklisted_fields
+                             blacklisted_fields: @blacklisted_fields,
+                             export_date: @notifier.export_date
```
Comment (gaorlov) [inline lib/aurora_bootstrapper/exporter.rb:28]: How does this handle the use case when the user doesn't want a date in the path?

Comment (YinfangZ) [reply lib/aurora_bootstrapper/exporter.rb:28]: why we do not want one? if there were no datetime subfolder, the data files would end up under e.g. backfill/
app1d/app1d-mysql1/

Comment (YinfangZ) [reply lib/aurora_bootstrapper/exporter.rb:28]: If two db dumps are on going it could overwrite each other, leaving the data files in inconsistent mode

Comment (gaorlov) [reply lib/aurora_bootstrapper/exporter.rb:28]: Correct. While this is not the usecase Data Plat wants, this is how BI expects data to be exported.

Comment (YinfangZ) [reply lib/aurora_bootstrapper/exporter.rb:28]: Then BI team needs to change their script logic. We cannot do if elif elif ... for different teams here. We define the folder structure.

File: lib/aurora_bootstrapper/exporter.rb [hunk 18]
```diff
@@ -11,23 +11,32 @@ def initialize( client:, prefix: "", export_bucket:, blacklisted_tables: "", whi
...
         rescue => e
           AuroraBootstrapper.logger.error message: "Error in database #{database_name}", error: e
         end
       end
+
+      notify!
+    end
+
+    def notify!
+      @notifier.notify
```
Comment (gaorlov) [inline lib/aurora_bootstrapper/exporter.rb:39]: what's the benefit of wrapping this in a method? Can we just call `@notifier.notify` from `export`?

**Review by gaorlov — COMMENTED** [review]
This looks great!
The only outstanding things are:
* What happens when we don't want to specify an export date?
* I don't see any tests for the notifier. Let's aim to have test coverage as close to 100% as we can.
Nice work!

File: lib/aurora_bootstrapper/exporter.rb (hunk 17 above)
Comment (YinfangZ) [reply lib/aurora_bootstrapper/exporter.rb:28]: If BI team does not want it it can explicitly set EXPORT_DATE to be empty.

File: lib/aurora_bootstrapper/exporter.rb (hunk 18 above)
Comment (YinfangZ) [reply lib/aurora_bootstrapper/exporter.rb:39]: For testing

Comment (gaorlov) [reply lib/aurora_bootstrapper/exporter.rb:39]: What is the test case that we're solving for? If the test is for the notifier, can it be addressed in the notifier suite?

Comment (YinfangZ) [reply lib/aurora_bootstrapper/exporter.rb:39]: Removed and found a way to test

File: lib/aurora_bootstrapper/exporter.rb (hunk 17 above)
Comment (gaorlov) [reply lib/aurora_bootstrapper/exporter.rb:28]: This introduces a breaking behavior change: previously omitting `EXPORT_DATE` would have skipped it, but now the user will have to add the date, but make it blank. Is this what you would expect as an end user?

Setting `EXPORT_DATE` to empty will evaluate to empty string.
```
> EXPORT_DATE= irb
2.5.0 :001 > ENV['EXPORT_DATE']
 => "" 
```
Will the code gracefully handle this case?

Comment (YinfangZ) [reply lib/aurora_bootstrapper/exporter.rb:28]: Add another env variable EXPORT_DATE_TURN_OFF

Comment (gaorlov) [reply lib/aurora_bootstrapper/exporter.rb:28]: Can you help me understand how this is not a breaking change?

Comment (YinfangZ) [reply lib/aurora_bootstrapper/exporter.rb:28]: Synced up offline.

File: test/lib/notifier_test.rb [hunk 19]
```diff
@@ -0,0 +1,67 @@
+require 'test_helper'
+
+class TableTest < Minitest::Test
```
Comment (gaorlov) [inline test/lib/notifier_test.rb:3]: this will override the existing `TableTest` class. Let's rename it to `NotifierTest`

File: test/lib/notifier_test.rb [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+    assert_equal export_date, AuroraBootstrapper::Notifier.new(s3_path: @bukkit).export_date
+  end
+
+  def test_region
+    region = ENV.fetch "REGION"
+    assert_equal region, AuroraBootstrapper::Notifier.new(s3_path: @bukkit).send(:region)
+  end
+
+  def test_unprefixed_path
+    assert_equal "bukkit", AuroraBootstrapper::Notifier.new(s3_path: @bukkit).send(:unprefixed_path)
```
Comment (gaorlov) [inline test/lib/notifier_test.rb:20]: can we test this with a longer path name? like `s3://bukkit/something/something`

File: test/lib/notifier_test.rb (hunk 19 above)
Comment (YinfangZ) [reply test/lib/notifier_test.rb:3]: eh copy and paste sin

File: test/lib/notifier_test.rb (hunk 20 above)
Comment (YinfangZ) [reply test/lib/notifier_test.rb:20]: Sure.

File: lib/aurora_bootstrapper/notifier.rb [hunk 21]
```diff
@@ -0,0 +1,55 @@
+module AuroraBootstrapper
+  class Notifier
+    def initialize( s3_path: s3_path)
+        @s3_path = s3_path
+        if ENV.fetch( 'EXPORT_DATE_OVERRIDE', false )
+          @export_date ||= ENV.fetch( 'EXPORT_DATE', DateTime.now.strftime("%Y-%m-%d") )
```
Comment (gaorlov) [inline lib/aurora_bootstrapper/notifier.rb:6]: does this mean that if we have `EXPORT_DATE_OVERRIDE` unset, we will not set `@export_date` regardless of the value of `EXPORT_DATE`?

Comment (YinfangZ) [reply lib/aurora_bootstrapper/notifier.rb:6]: yup

Comment (YinfangZ) [reply lib/aurora_bootstrapper/notifier.rb:6]: It will affect BI team

Comment (YinfangZ) [reply lib/aurora_bootstrapper/notifier.rb:6]: what could be the case when EXPORT_DATE_OVERRIDE is unset and EXPORT_DATE is set?

Comment (YinfangZ) [reply lib/aurora_bootstrapper/notifier.rb:6]: We should remove EXPORT_DATE altogether and only use EXPORT_DATE_OVERRIDE and current day. The logic will be simpler to follow. There is no reason to put data files under datetime folder other than NOW

Comment (gaorlov) [reply lib/aurora_bootstrapper/notifier.rb:6]: So just to make sure I understand
The options are:
```sh
# EXPORT_DATE_OVERRIDE unset
# EXPORT_DATE unset
```
no export date will be set.

```sh
export EXPORT_DATE_OVERRIDE=true
export EXPORT_DATE=date
```
export date set to the var

```bash
export EXPORT_DATE_OVERRIDE=true
# EXPORT_DATE unset
```
will auto set the date

```bash
# EXPORT_DATE_OVERRIDE unset
export EXPORT_DATE=date
```
__no export date will be set__
Is this expected?

Comment (YinfangZ) [reply lib/aurora_bootstrapper/notifier.rb:6]: Theoretically this should never happen; we could throw exception

Comment (YinfangZ) [reply lib/aurora_bootstrapper/notifier.rb:6]: If OVERRIDE flag is set and some user specifies EXPORT_DATE, he/she should have get Error

Comment (YinfangZ) [reply lib/aurora_bootstrapper/notifier.rb:6]: OVERRIDE flag shows intention of using datetime folder

Comment (YinfangZ) [reply lib/aurora_bootstrapper/notifier.rb:6]: Or you want to be backward compatible??

Comment (gaorlov) [reply lib/aurora_bootstrapper/notifier.rb:6]: basck compat is ideal

Comment (YinfangZ) [reply lib/aurora_bootstrapper/notifier.rb:6]: I already updated as you said. Count you check line 8?

Comment (gaorlov) [reply lib/aurora_bootstrapper/notifier.rb:6]: would this be easier to follow if it were
```ruby
def export_date
  @export_date ||= ENV.fetch('EXPORT_DATE', export_date_override )
end

def export_date_override
  DateTime.now.strftime("%Y-%m-%d") if ENV.[ 'EXPORT_DATE_OVERRIDE' ]
end
```

Comment (YinfangZ) [reply lib/aurora_bootstrapper/notifier.rb]: Done

File: bin/exporter (hunk 9 above)
Comment (gaorlov) [inline bin/exporter:32]: looks like this gets overridden in the `Exporter` initializer. Can we remove this?

File: lib/aurora_bootstrapper/notifier.rb [hunk 22]
```diff
@@ -0,0 +1,57 @@
+module AuroraBootstrapper
+  class Notifier
+    def initialize( s3_path: s3_path)
+        @s3_path = s3_path
+        export_date
```
Comment (gaorlov) [inline lib/aurora_bootstrapper/notifier.rb:5]: total nit: we don't need to call this method here

File: test/lib/notifier_test.rb [hunk 23]
```diff
@@ -0,0 +1,66 @@
+require 'test_helper'
+
+class NotifierTest < Minitest::Test
+  def setup
+    @bukkit               = ENV.fetch "EXPORT_BUCKET"
+  end
+
+  def test_export_date
+    export_date = ENV.fetch "EXPORT_DATE"
+    assert_equal export_date, AuroraBootstrapper::Notifier.new(s3_path: @bukkit).export_date
```
Comment (gaorlov) [inline test/lib/notifier_test.rb:10]: should we test the override logic here? It's a bit involved and it would be nice to have the peace of mind that it's covered in case we ever need to make changes

**Review by gaorlov — COMMENTED** [review]
Looks good!
Let's get that last bit of test coverage on the date overrides and we'll be gtg. Thanks so much of doing this!

File: test/lib/notifier_test.rb (hunk 23 above)
Comment (YinfangZ) [reply test/lib/notifier_test.rb:10]: yup no surprises

**Review by gaorlov — APPROVED** [review]
LGTM
I'll bump the version and push it up once this goes into master. 
Great work!
